logstash-core/spec/logstash/java_pipeline_spec.rb

```diff
@@ -260,6 +283,61 @@ def flush(options)
     end
   end
 
+  context "when the output plugin raises an abort batch exception" do
+    let(:metric) { LogStash::Instrument::Metric.new(LogStash::Instrument::Collector.new) }
+    subject { mock_java_pipeline_from_string(config, pipeline_settings_obj, metric) }
+    let(:metric_store) { subject.metric.collector.snapshot_metric.metric_store }
+    let(:config) do
+      <<-EOS
+      input { dummymanualinputgenerator {iterations => 2} }
+      output { dummyabortingoutput {} }
+      EOS
+    end
+    let(:dummyabortingoutput) { DummyAbortingOutput.new }
+    let(:pipeline_settings) { { "pipeline.batch.size" => 2, "queue.type" => "persisted"} }
+
+    let(:collected_metric) { metric_store.get_with_path("stats/events") }
+    let (:queue_client_batch) { double("Acked queue client Mock") }
+    let(:logger) { double("pipeline logger").as_null_object }
+
+    before :each do
+      # warn: use a real DummyAbortingOutput plugin instantiated by PluginFactory because it needs
+      # a properly initialized ExecutionContext, so DONT' mock the constructor of DummyAbortingOutput
+      allow(LogStash::Plugin).to receive(:lookup).with("input", "dummymanualinputgenerator").and_return(DummyManualInputGenerator)
+      allow(LogStash::Plugin).to receive(:lookup).with("codec", "plain").and_return(LogStash::Codecs::Plain)
+      allow(LogStash::Plugin).to receive(:lookup).with("output", "dummyabortingoutput").and_return(DummyAbortingOutput)
+      allow(LogStash::JavaPipeline).to receive(:filter_queue_client).and_return(queue_client_batch)
+      allow(LogStash::JavaPipeline).to receive(:logger).twice.and_return(logger)
+    end
+
+    it "should not acknowledge the batch" do
+      expect { subject.start }.to_not raise_error
+      expect(queue_client_batch).to_not receive(:close_batch)
+      expect(logger).not_to receive(:error).with(/Pipeline worker error, the pipeline will be stopped/, anything)
+
+      # make sure all the workers are started
+      wait(5).for {subject.worker_threads.any?(&:alive?)}.to be_truthy
+
+      # command a shutdown while the output is processing a batch and not completing it
+      thread = Thread.new { subject.shutdown_workers }
+
+      # wait for inputs to terminate
+      wait(5).for {subject.input_threads.any?(&:alive?)}.to be_falsey
+
+      # the exception raised by the aborting output should have stopped the workers
+      wait(5).for {subject.worker_threads.any?(&:alive?)}.to be_falsey
+
+      # need to wait that the pipeline thread stops completely, else the logger mock could be
+      # used outside of this context
+      subject.wait_for_shutdown
+
+      # verify the not completed batch by checking output stats
+      expect(collected_metric[:stats][:events][:duration_in_millis].value).not_to be_nil
+      expect(collected_metric[:stats][:events][:in].value).to eq(2)
+      expect(collected_metric[:stats][:events][:out].value).to eq(0)
+    end
```

Comment:
I suggest separating these into two tests:
1. that confirms that we don't ack the back
2. that shows we don't explode unpredictably

```suggestion
   before(:each) do
      expect { subject.start }.to_not raise_error
      expect(queue_client_batch).to_not receive(:close_batch)

      # make sure all the workers are started
      wait(5).for {subject.worker_threads.any?(&:alive?)}.to be_truthy
    end
    it "should not acknowledge the batch" do
      # command a shutdown while the output is processing a batch and not completing it
      thread = Thread.new { subject.shutdown_workers }

      # wait for inputs to terminate
      wait(5).for {subject.input_threads.any?(&:alive?)}.to be_falsey

      # the exception raised by the aborting output should have stopped the workers
      wait(5).for {subject.worker_threads.any?(&:alive?)}.to be_falsey

      # need to wait that the pipeline thread stops completely, else the logger mock could be
      # used outside of this context
      subject.wait_for_shutdown

      # verify the not completed batch by checking output stats
      expect(collected_metric[:stats][:events][:duration_in_millis].value).not_to be_nil
      expect(collected_metric[:stats][:events][:in].value).to eq(2)
      expect(collected_metric[:stats][:events][:out].value).to eq(0)
    end

    it "should not throw a generic error" do
      expect(logger).not_to receive(:error).with(/Pipeline worker error, the pipeline will be stopped/, anything)

      # command a shutdown while the output is processing a batch and not completing it
      thread = Thread.new { subject.shutdown_workers }

      # wait for inputs to terminate
      wait(5).for {subject.input_threads.any?(&:alive?)}.to be_falsey

      # the exception raised by the aborting output should have stopped the workers
      wait(5).for {subject.worker_threads.any?(&:alive?)}.to be_falsey

      # need to wait that the pipeline thread stops completely, else the logger mock could be
      # used outside of this context
      subject.wait_for_shutdown
    end
```